Remote-dev with mutable jar throws connection errors

**Describe the bug**
Since the change in how the `remote-dev` feature works (was that around `1.6`, `1.7`?), the remote dev doesn't quite work for me anymore.

Before, I've used this approach: https://blog.sebastian-daschner.com/entries/quarkus-remote-dev-in-containers

Now, I've tried the following, with `mutable-jar` and `QUARKUS_LAUNCH_DEVMODE=true` on the production side, which *basically* works, but results in many connection errors, and also reacts much slower than usual, since it involves some restarts (?).

**To Reproduce**
- The following branch: https://github.com/sdaschner/quarkus-playground/tree/remote-dev
- `mvn clean package`
- `export QUARKUS_LAUNCH_DEVMODE=true`
- `java -jar target/quarkus-app/quarkus-run.jar`
- In another terminal: `mvn quarkus:remote-dev -Dquarkus.live-reload.url=http://localhost:8080`
- In another terminal: `curl localhost:8080/coffee`

When I change some Java code and make another HTTP request, it produces the following errors:

In `mvn`:

```
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------< com.sebastian-daschner:quarkus-playground >--------------
[INFO] Building quarkus-playground 1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:1.8.1.Final:remote-dev (default-cli) @ quarkus-playground ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 10 source files to /home/sebastian/workspaces/sdaschner/quarkus-playground/target/classes
Listening for transport dt_socket at address: 5005
2020-10-03 10:14:17,646 INFO  [org.jbo.threads] (main) JBoss Threads version 3.1.1.Final
2020-10-03 10:14:19,901 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 2997ms
2020-10-03 10:14:20,396 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Connected to remote server
2020-10-03 10:14:20,590 ERROR [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Remote dev request failed: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:503)
	at java.base/sun.nio.ch.Net.connect(Net.java:492)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:588)
	at java.base/java.net.Socket.connect(Socket.java:648)
	at java.base/java.net.Socket.connect(Socket.java:597)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:182)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:859)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:689)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1618)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1523)
	at io.quarkus.vertx.http.deployment.devmode.HttpRemoteDevClient$Session.run(HttpRemoteDevClient.java:188)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:14:22,593 ERROR [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Remote dev request failed: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:503)
	at java.base/sun.nio.ch.Net.connect(Net.java:492)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:588)
	at java.base/java.net.Socket.connect(Socket.java:648)
	at java.base/java.net.Socket.connect(Socket.java:597)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:182)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
	at java.base/sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:341)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:362)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1256)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1194)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1082)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1016)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1370)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1345)
	at io.quarkus.vertx.http.deployment.devmode.HttpRemoteDevClient$Session.run(HttpRemoteDevClient.java:186)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:14:24,735 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Connected to remote server
2020-10-03 10:16:42,361 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (Remote dev client thread) Changed source files detected, recompiling [/home/sebastian/workspaces/sdaschner/quarkus-playground/src/main/java/com/sebastian_daschner/coffee/CoffeeShop.java]
2020-10-03 10:16:44,755 INFO  [io.qua.dep.QuarkusAugmentor] (Remote dev client thread) Quarkus augmentation completed in 801ms
2020-10-03 10:16:44,768 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (Remote dev client thread) Hot replace total time: 2.409s 
2020-10-03 10:16:44,848 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending app/quarkus-playground.jar
2020-10-03 10:16:44,852 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending dev/app/com/sebastian_daschner/coffee/CoffeeShop.class
2020-10-03 10:16:44,855 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending quarkus-run.jar
2020-10-03 10:16:44,859 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending lib/deployment/build-system.properties
2020-10-03 10:16:45,297 ERROR [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Remote dev request failed: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:323)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:350)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:803)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:982)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:244)
	at java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:343)
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:754)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:689)
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:863)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:689)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1618)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1523)
	at io.quarkus.vertx.http.deployment.devmode.HttpRemoteDevClient$Session.run(HttpRemoteDevClient.java:188)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:16:47,299 ERROR [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Remote dev request failed: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:503)
	at java.base/sun.nio.ch.Net.connect(Net.java:492)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:588)
	at java.base/java.net.Socket.connect(Socket.java:648)
	at java.base/java.net.Socket.connect(Socket.java:597)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:182)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
	at java.base/sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:341)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:362)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1256)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1194)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1082)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1016)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1370)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1345)
	at io.quarkus.vertx.http.deployment.devmode.HttpRemoteDevClient$Session.run(HttpRemoteDevClient.java:186)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:16:49,449 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending app/quarkus-playground.jar
2020-10-03 10:16:49,452 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending dev/app/com/sebastian_daschner/coffee/CoffeeShop.class
2020-10-03 10:16:49,454 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending quarkus-run.jar
2020-10-03 10:16:49,456 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Sending lib/deployment/build-system.properties
2020-10-03 10:16:49,458 INFO  [io.qua.ver.htt.dep.dev.HttpRemoteDevClient] (Remote dev client thread) Connected to remote server
```

In `java -jar`:

```
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-03 10:13:18,360 INFO  [io.qua.sch.run.SimpleScheduler] (Quarkus Main Thread) No scheduled business methods found - Simple scheduler will not be started
2020-10-03 10:13:18,445 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground 1.0 on JVM (powered by Quarkus 1.8.1.Final) started in 2.234s. Listening on: http://0.0.0.0:8080
2020-10-03 10:13:18,446 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-10-03 10:13:18,447 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy, resteasy-jsonb, scheduler, smallrye-health]
2020-10-03 10:14:20,559 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground stopped in 0.085s
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-03 10:14:23,140 INFO  [io.qua.sch.run.SimpleScheduler] (Quarkus Main Thread) No scheduled business methods found - Simple scheduler will not be started
2020-10-03 10:14:23,203 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground 1.0 on JVM (powered by Quarkus 1.8.1.Final) started in 2.148s. Listening on: http://0.0.0.0:8080
2020-10-03 10:14:23,203 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-10-03 10:14:23,203 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy, resteasy-jsonb, scheduler, smallrye-health]
2020-10-03 10:14:24,663 ERROR [io.qua.ver.htt.run.dev.RemoteSyncHandler] (vert.x-eventloop-thread-2) Invalid session
2020-10-03 10:16:44,916 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground stopped in 0.051s
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-03 10:16:45,667 ERROR [io.qua.run.boo.ForkJoinClassLoading] (vert.x-worker-thread-1) Failed to set fork join ClassLoader: java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1080)
	at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:278)
	at io.quarkus.runner.bootstrap.ForkJoinClassLoading.setForkJoinClassLoader(ForkJoinClassLoading.java:40)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:117)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:176)
	at io.quarkus.deployment.dev.IsolatedDevModeMain$$Lambda$118/0000000000000000.accept(Unknown Source)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:183)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:61)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:51)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.ContextImpl$$Lambda$729/0000000000000000.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:16:45,813 ERROR [io.qua.application] (Quarkus Main Thread) Failed to start application (with profile dev): java.util.concurrent.RejectedExecutionException: Task io.vertx.core.impl.TaskQueue$$Lambda$675/0000000000000000@49d13145 rejected from java.util.concurrent.ThreadPoolExecutor@5c3e938f[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2057)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:827)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1357)
	at io.vertx.core.impl.TaskQueue.execute(TaskQueue.java:93)
	at io.vertx.core.impl.ContextImpl.executeBlocking(ContextImpl.java:334)
	at io.vertx.core.impl.ContextImpl.executeBlocking(ContextImpl.java:283)
	at io.vertx.core.impl.EventLoopContext.executeBlocking(EventLoopContext.java:23)
	at io.vertx.core.impl.VertxImpl.executeBlocking(VertxImpl.java:748)
	at io.vertx.core.impl.VertxImpl.executeBlocking(VertxImpl.java:754)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder.tryCleanTccl(VertxCoreRecorder.java:115)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder.configureVertx(VertxCoreRecorder.java:77)
	at io.quarkus.deployment.steps.VertxCoreProcessor$build-956362597.deploy_0(VertxCoreProcessor$build-956362597.zig:117)
	at io.quarkus.deployment.steps.VertxCoreProcessor$build-956362597.deploy(VertxCoreProcessor$build-956362597.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:468)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:91)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
	at java.base/java.lang.Thread.run(Thread.java:835)

2020-10-03 10:16:45,816 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-1) Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:176)
	at io.quarkus.deployment.dev.IsolatedDevModeMain$$Lambda$118/0000000000000000.accept(Unknown Source)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:183)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:61)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:51)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.ContextImpl$$Lambda$729/0000000000000000.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:717)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:91)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
	... 1 more
Caused by: java.util.concurrent.RejectedExecutionException: Task io.vertx.core.impl.TaskQueue$$Lambda$675/0000000000000000@49d13145 rejected from java.util.concurrent.ThreadPoolExecutor@5c3e938f[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2057)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:827)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1357)
	at io.vertx.core.impl.TaskQueue.execute(TaskQueue.java:93)
	at io.vertx.core.impl.ContextImpl.executeBlocking(ContextImpl.java:334)
	at io.vertx.core.impl.ContextImpl.executeBlocking(ContextImpl.java:283)
	at io.vertx.core.impl.EventLoopContext.executeBlocking(EventLoopContext.java:23)
	at io.vertx.core.impl.VertxImpl.executeBlocking(VertxImpl.java:748)
	at io.vertx.core.impl.VertxImpl.executeBlocking(VertxImpl.java:754)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder.tryCleanTccl(VertxCoreRecorder.java:115)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder.configureVertx(VertxCoreRecorder.java:77)
	at io.quarkus.deployment.steps.VertxCoreProcessor$build-956362597.deploy_0(VertxCoreProcessor$build-956362597.zig:117)
	at io.quarkus.deployment.steps.VertxCoreProcessor$build-956362597.deploy(VertxCoreProcessor$build-956362597.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:468)
	... 12 more

2020-10-03 10:16:45,833 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-1) Hot replace total time: 3.482s 
2020-10-03 10:16:46,377 INFO  [org.jbo.threads] (Classpath Change Timer) JBoss Threads version 3.1.1.Final
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-03 10:16:48,280 INFO  [io.qua.sch.run.SimpleScheduler] (Quarkus Main Thread) No scheduled business methods found - Simple scheduler will not be started
2020-10-03 10:16:48,355 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground 1.0 on JVM (powered by Quarkus 1.8.1.Final) started in 3.423s. Listening on: http://0.0.0.0:8080
2020-10-03 10:16:48,355 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-10-03 10:16:48,355 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy, resteasy-jsonb, scheduler, smallrye-health]
2020-10-03 10:16:49,377 ERROR [io.qua.ver.htt.run.dev.RemoteSyncHandler] (vert.x-eventloop-thread-2) Invalid session
2020-10-03 10:16:51,925 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground stopped in 0.025s
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-03 10:16:52,439 INFO  [io.qua.sch.run.SimpleScheduler] (Quarkus Main Thread) No scheduled business methods found - Simple scheduler will not be started
2020-10-03 10:16:52,440 INFO  [io.quarkus] (Quarkus Main Thread) quarkus-playground 1.0 on JVM (powered by Quarkus 1.8.1.Final) started in 0.511s. Listening on: http://0.0.0.0:8080
2020-10-03 10:16:52,440 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-10-03 10:16:52,440 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy, resteasy-jsonb, scheduler, smallrye-health]
2020-10-03 10:16:52,441 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-0) Hot replace total time: 0.552s
```

I've used:

- Quarkus: `1.8.1.Final`
- Java: `Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.19.0, JRE 14 Linux amd64-64-Bit Compressed References 20200318_6 (JIT enabled, AOT enabled)`

With that experience, I'm blocked to update to the latest version. What is the correct way to run this new remote-dev mode?

--- a/extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
+++ b/extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
@@ -9,6 +9,7 @@ import static io.vertx.core.file.impl.FileResolver.CACHE_DIR_BASE_PROP_NAME;
 
 import java.io.File;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
@@ -372,6 +373,11 @@ public class VertxCoreRecorder {
                 return threads;
             }
         };
+    }
+
+    public static Supplier<Vertx> recoverFailedStart(VertxConfiguration config) {
+        return vertx = new VertxSupplier(config, Collections.emptyList());
+
     }
 
     static class VertxSupplier implements Supplier<Vertx> {

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/HttpRemoteDevClient.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/HttpRemoteDevClient.java
@@ -65,6 +65,7 @@ public class HttpRemoteDevClient implements RemoteDevClient {
         private final Thread httpThread;
         private final String url;
         private final URL devUrl;
+        private final URL probeUrl;
         int errorCount;
 
         private Session(RemoteDevState initialState,
@@ -74,6 +75,7 @@ public class HttpRemoteDevClient implements RemoteDevClient {
             this.initialConnectFunction = initialConnectFunction;
             this.changeRequestFunction = changeRequestFunction;
             devUrl = new URL(HttpRemoteDevClient.this.url + RemoteSyncHandler.DEV);
+            probeUrl = new URL(HttpRemoteDevClient.this.url + RemoteSyncHandler.PROBE);
             url = HttpRemoteDevClient.this.url;
             httpThread = new Thread(this, "Remote dev client thread");
             httpThread.start();
@@ -248,7 +250,9 @@ public class HttpRemoteDevClient implements RemoteDevClient {
             }
             while (System.currentTimeMillis() < timeout) {
                 try {
-                    HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+                    HttpURLConnection connection = (HttpURLConnection) probeUrl.openConnection();
+                    connection.setRequestMethod("POST");
+                    connection.addRequestProperty(HttpHeaders.CONTENT_TYPE.toString(), RemoteSyncHandler.APPLICATION_QUARKUS);
                     IoUtil.readBytes(connection.getInputStream());
                     return doConnect(initialState, initialConnectFunction);
                 } catch (IOException e) {

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -173,7 +173,7 @@ public class VertxHttpRecorder {
         if (supplier == null) {
             VertxConfiguration vertxConfiguration = new VertxConfiguration();
             ConfigInstantiator.handleObject(vertxConfiguration);
-            vertx = VertxCoreRecorder.initialize(vertxConfiguration, null);
+            vertx = VertxCoreRecorder.recoverFailedStart(vertxConfiguration).get();
         } else {
             vertx = supplier.get();
         }
@@ -187,7 +187,14 @@ public class VertxHttpRecorder {
             if (hotReplacementHandler != null) {
                 router.route().order(Integer.MIN_VALUE).blockingHandler(hotReplacementHandler);
             }
-            rootHandler = router;
+            Handler<HttpServerRequest> root = router;
+            LiveReloadConfig liveReloadConfig = new LiveReloadConfig();
+            ConfigInstantiator.handleObject(liveReloadConfig);
+            if (liveReloadConfig.password.isPresent()
+                    && hotReplacementContext.getDevModeType() == DevModeType.REMOTE_SERVER_SIDE) {
+                root = remoteSyncHandler = new RemoteSyncHandler(liveReloadConfig.password.get(), root, hotReplacementContext);
+            }
+            rootHandler = root;
 
             //we can't really do
             doServerStart(vertx, buildConfig, config, LaunchMode.DEVELOPMENT, new Supplier<Integer>() {

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/RemoteSyncHandler.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/RemoteSyncHandler.java
@@ -14,9 +14,10 @@ import org.jboss.logging.Logger;
 import io.netty.handler.codec.http.HttpHeaderNames;
 import io.quarkus.dev.spi.HotReplacementContext;
 import io.quarkus.dev.spi.RemoteDevState;
-import io.quarkus.runtime.ExecutorRecorder;
 import io.quarkus.runtime.util.HashUtil;
+import io.quarkus.vertx.core.runtime.VertxCoreRecorder;
 import io.vertx.core.Handler;
+import io.vertx.core.Promise;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpMethod;
 import io.vertx.core.http.HttpServerRequest;
@@ -31,6 +32,7 @@ public class RemoteSyncHandler implements Handler<HttpServerRequest> {
     public static final String QUARKUS_SESSION_COUNT = "X-Quarkus-Count";
     public static final String CONNECT = "/connect";
     public static final String DEV = "/dev";
+    public static final String PROBE = "/probe"; //used to check that the server is back up after restart
 
     final String password;
     final Handler<HttpServerRequest> next;
@@ -77,12 +79,16 @@ public class RemoteSyncHandler implements Handler<HttpServerRequest> {
         final String type = event.headers().get(HttpHeaderNames.CONTENT_TYPE);
         if (APPLICATION_QUARKUS.equals(type)) {
             currentSessionTimeout = time + 60000;
-            ExecutorRecorder.getCurrent().execute(new Runnable() {
+            VertxCoreRecorder.getVertx().get().executeBlocking(new Handler<Promise<Object>>() {
                 @Override
-                public void run() {
-                    handleRequest(event);
+                public void handle(Promise<Object> promise) {
+                    try {
+                        handleRequest(event);
+                    } finally {
+                        promise.complete();
+                    }
                 }
-            });
+            }, null);
             return;
         }
         next.handle(event);
@@ -98,6 +104,8 @@ public class RemoteSyncHandler implements Handler<HttpServerRequest> {
                 handleDev(event);
             } else if (event.path().equals(CONNECT)) {
                 handleConnect(event);
+            } else if (event.path().equals(PROBE)) {
+                event.response().end();
             } else {
                 event.response().setStatusCode(404).end();
             }
@@ -114,9 +122,9 @@ public class RemoteSyncHandler implements Handler<HttpServerRequest> {
                 if (checkSession(event, b.getBytes())) {
                     return;
                 }
-                ExecutorRecorder.getCurrent().execute(new Runnable() {
+                VertxCoreRecorder.getVertx().get().executeBlocking(new Handler<Promise<Object>>() {
                     @Override
-                    public void run() {
+                    public void handle(Promise<Object> promise) {
                         try {
                             Throwable problem = (Throwable) new ObjectInputStream(new ByteArrayInputStream(b.getBytes()))
                                     .readObject();
@@ -144,10 +152,11 @@ public class RemoteSyncHandler implements Handler<HttpServerRequest> {
                         } catch (Exception e) {
                             log.error("Connect failed", e);
                             event.response().setStatusCode(500).end();
+                        } finally {
+                            promise.complete();
                         }
                     }
-                });
-
+                }, null);
             }
         }).exceptionHandler(new Handler<Throwable>() {
             @Override

